009-quarkus-infinispan-grpc-kafka/src/test/java/io/quarkus/qe/grateful/shutdown/KafkaGratefulShutdownTest.java

```diff
@@ -0,0 +1,122 @@
+
+public class KafkaGratefulShutdownTest {
+
+    private static final int TOTAL_MESSAGES = 10;
+    private static final String GRATEFUL_SHUTDOWN_PROPERTY = "mp.messaging.incoming.slow.graceful-shutdown";
+    private static final String KAFKA_LOG_PROPERTY = "quarkus.log.category.\"io.smallrye.reactive.messaging.kafka\".level";
+    private static final String LAST_MESSAGE_LOG = "Processed Message " + TOTAL_MESSAGES;
+    private static final String SHUTDOWN_KAFKA_LOG = "Shutting down - Waiting for message processing to complete";
+
+    @RegisterExtension
+    static QuarkusProdModeTest app = new QuarkusProdModeTest()
+            .setArchiveProducer(
+                    () -> ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(SlowTopicResource.class, SlowTopicConsumer.class))
+            .setRuntimeProperties(Collections.singletonMap("quarkus.http.port", "8083"))
+            .setLogFileName("out.log")
+            .overrideConfigKey(KAFKA_LOG_PROPERTY, Level.DEBUG.getName())
+            .setRun(true);
+
+    @LogFile
+    Path logfile;
+
+    @Test
+    public void shouldWaitForMessagesWhenGratefulShutdownIsEnabled() throws InterruptedException {
+        givenApplicationWithGratefulShutdownEnabled();
+        givenMessagesInTopic();
+        whenStopApplication();
+        thenAllMessagesAreProcessedOrKafkaIsShutdown();
+    }
+
+    @Test
+    public void shouldNotWaitForMessagesWhenGratefulShutdownIsDisabled() throws InterruptedException {
+        givenApplicationWithGratefulShutdownDisabled();
+        givenMessagesInTopic();
+        whenStopApplication();
+        thenAllMessagesAreNotProcessed();
+    }
+
+    private void givenApplicationWithGratefulShutdownEnabled() {
+        app.overrideConfigKey(GRATEFUL_SHUTDOWN_PROPERTY, Boolean.TRUE.toString());
+        restart();
+    }
+
+    private void givenApplicationWithGratefulShutdownDisabled() {
+        app.overrideConfigKey(GRATEFUL_SHUTDOWN_PROPERTY, Boolean.FALSE.toString());
+        restart();
+    }
+
+    private void givenMessagesInTopic() {
+        post("/slow-topic/sendMessages/" + TOTAL_MESSAGES);
+    }
+
+    private void whenStopApplication() {
+        app.stop();
+    }
+
+    private void thenAllMessagesAreProcessedOrKafkaIsShutdown() {
+        thenAssertLogs(line -> line.contains(LAST_MESSAGE_LOG) || line.contains(SHUTDOWN_KAFKA_LOG),
+                "Expected output was not found in logs");
+    }
+
+    private void thenAllMessagesAreNotProcessed() {
+        thenLogsDoNotContain(LAST_MESSAGE_LOG);
+        thenLogsDoNotContain(SHUTDOWN_KAFKA_LOG);
+    }
+
+    private void thenLogsDoNotContain(String expectedOutput) {
+        thenAssertLogs(line -> !line.contains(expectedOutput), "Unexpected output was found in logs");
+    }
+
+    private void thenAssertLogs(Predicate<String> assertion, String message) {
+        // the logs might not be flushed to disk immediately, so wait a few seconds before giving up completely
```

Comment:
I don't think that this, could be an issue.

Reply:
I copied and pasted this comment, will remove it.